### PR TITLE
[SPARK-12371][SQL] Runtime nullability check for NewInstance

### PR DESCRIPTION
**Pull request by liancheng**
This PR adds a new expression `AssertNotNull` to ensure non-nullable fields of products and case classes don't receive null values at runtime.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -49,6 +49,7 @@ import org.apache.spark.sql.types._
  *
  */
 abstract class Expression extends TreeNode[Expression] {
+  protected def nullabilitySuffix: String = if (nullable) "<?>" else "<!>"
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: This is used to show nullability in the query plan / expression tree. I found it useful while debugging nullability issues.

**Comment by liancheng** [comment]
Hm, similar as #10296, this PR also caught several existing nullability inconsistency bugs, which caused the last build failure. Trying to fix them.

**Comment by yhuai** [comment]
@liancheng Regarding the scope of this jira, my understanding is that when we create JVM objects, if there is any null values and we are trying to set them to primitive fields of the class, we should throw a Runtime exception to ask them to use Option or non-primitive JVM type (e.g. java.lang.Integer instead of Int). This is a runtime check. For example, we want to encode a row to a case class `case class TestData(a: Int)`. If the first value of a row is `null`, we should throw the Runtime exception when we try to create a `TestData` instance because `a` is a primitive type field.

**Comment by liancheng** [comment]
@yhuai Thanks a lot for the explanation, I misunderstood the scope of the JIRA ticket. Updated this PR according to @marmbrus's [comment][1] in #10296. A new expression `AssertNotNull` is added to assert non-nullable constructor arguments are indeed non-null.

**Comment by liancheng** [comment]
retest this please

File: sql/catalyst/src/test/resources/log4j.properties [hunk 2]
```diff
@@ -16,9 +16,9 @@
 #
 
 # Set everything to be logged to the file target/unit-tests.log
-log4j.rootCategory=INFO, file
+log4j.rootCategory=TRACE, file
 log4j.appender.file=org.apache.log4j.FileAppender
-log4j.appender.file.append=true
+log4j.appender.file.append=false
```
Comment (yhuai) [inline sql/catalyst/src/test/resources/log4j.properties]: remove these?

File: sql/core/src/test/resources/log4j.properties [hunk 3]
```diff
@@ -33,7 +33,7 @@ log4j.appender.FA.layout=org.apache.log4j.PatternLayout
 log4j.appender.FA.layout.ConversionPattern=%d{HH:mm:ss.SSS} %t %p %c{1}: %m%n
 
 # Set the logger level of File Appender to WARN
-log4j.appender.FA.Threshold = INFO
+log4j.appender.FA.Threshold = TRACE
```
Comment (yhuai) [inline sql/core/src/test/resources/log4j.properties]: Revert these?

**Comment by yhuai** [comment]
Should we just keep the runtime part of changes?

**Comment by liancheng** [comment]
@yhuai Do you think we should move analysis phase checking into another PR or just drop that part? This check does find other nullability bugs (revealed by the Jenkins build failure). And I think Dataset nullability of Dataset schema should conforms to the underlying logical plan.

File: sql/catalyst/src/test/resources/log4j.properties (hunk 2 above)
Comment (liancheng) [inline sql/catalyst/src/test/resources/log4j.properties]: Oh yeah, thanks!

**Comment by yhuai** [comment]
Yeah, let's use this PR for the runtime check.

**Comment by liancheng** [comment]
@yhuai Narrowed down the scope of this PR. As we discussed offline, will open another one for the analysis phase check.

**Comment by liancheng** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 4]
```diff
@@ -20,17 +20,18 @@ package org.apache.spark.sql.catalyst.encoders
...
 import org.apache.spark.sql.catalyst.expressions.codegen.{GenerateSafeProjection, GenerateUnsafeProjection}
 import org.apache.spark.sql.catalyst.optimizer.SimplifyCasts
-import org.apache.spark.sql.catalyst.{JavaTypeInference, InternalRow, ScalaReflection}
-import org.apache.spark.sql.types.{StructField, ObjectType, StructType}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, Project}
+import org.apache.spark.sql.catalyst.util.sideBySide
+import org.apache.spark.sql.catalyst.{InternalRow, JavaTypeInference, ScalaReflection}
+import org.apache.spark.sql.types.{ObjectType, StructField, StructType}
+import org.apache.spark.sql.{AnalysisException, Encoder}
+import org.apache.spark.util.Utils
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Maybe we should revert the changes of this file?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -83,6 +84,16 @@ class Dataset[T] private[sql](
...
 
+  logTrace(
+    s"""
+       |# unresolvedTEncoder.fromRowExpression
+       |${unresolvedTEncoder.fromRowExpression.treeString}
+       |# resolvedTEncoder.fromRowExpression
+       |${resolvedTEncoder.fromRowExpression.treeString}
+       |# boundTEncoder.fromRowExpression
+       |${boundTEncoder.fromRowExpression.treeString}
+     """.stripMargin)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: What does this do?

**Comment by marmbrus** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 6]
```diff
@@ -515,12 +516,44 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+      val rowRDD = sqlContext.sparkContext.parallelize(rows)
+      sqlContext.createDataFrame(rowRDD, schema).as[NestedStruct]
+    }
+
+    checkAnswer(
+      buildDataset(Row(Row("hello", 1))),
+      NestedStruct(ClassData("hello", 1))
+    )
+
+    // Shouldn't throw runtime exception when parent object (`ClassData`) is null
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: My concern is: if the parent is null, we should shortcut the execution to return null directly, instead of going into the field and trigger the null check. However, looks like we only do this shortcut for product type by `If(IsNull...)`, we may also need to handle array type and map type.

Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Good catch! Verified that the current mechanism doesn't play well with primitive arrays.

Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Actually what the test case I constructed reflected is another separate bug, which has been fixed in PR #10401.

Discussed with @cloud-fan offline. What he meant was, we should also add `AssertNotNull` for array types and map types, with which I totally agree, but I think it would be nice to be added in a separate PR.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
The last build failure was irrelevant.

**Comment by liancheng** [comment]
Merging to master.
